src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1006,12 +1009,45 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
             this.SolutionState.WithOptions(options));
     }
 
+    public Task<SolutionCompilationState> WithFrozenPartialCompilationsAsync(CancellationToken cancellationToken)
+        => _cachedFrozenSnapshot.GetValueAsync(cancellationToken);
+
+    private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancellationToken)
+    {
+        var newIdToProjectStateMap = this.SolutionState.ProjectStates;
+        var newIdToTrackerMap = _projectIdToTrackerMap;
+
+        foreach (var (projectId, projectState) in newIdToProjectStateMap)
+        {
+            // if we don't have one or it is stale, create a new partial solution
+            var tracker = GetCompilationTracker(projectId);
+            var newTracker = tracker.FreezePartialState(this, cancellationToken);
+
+            Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(projectId));
+            newIdToProjectStateMap = newIdToProjectStateMap.SetItem(projectId, newTracker.ProjectState);
```

Comment:
Did you want to do an ImmutableDictionaryBuilder to avoid the allocations here?

Reply:
sure.